show /home based on survey status from correct, show content in /prog…

eilivjl opened the pull request:
…ress based on statuses from userStatuses

Use new structure of statuses in both firebase and in the app to display data on pages /home and /progress based on the progress of the respondent

TODO: add updateUserStatus (from another branch) to /progress to be able to set surveyStatus to COMPLETE in firebase

prabu-ssb commented on src/pages/progress.tsx:
```diff
@@ -49,7 +35,7 @@ const Progress = () => {
                     </div>
                 </div>
-                {purchases && purchases.length > 0 && userInfo.diaryStatus &&
+                {userStatuses && userStatuses.journalStatus === StatusConstants.COMPLETE &&
```
How about collapsing all these similar status checks into a utility function which does this --

```
export const isUserStatusComplete = (statusKey: UserStatusKeys): boolean => {
  const {userStatuses} = useContext(UserContext);
  return userStatuses[statusKey] === StatusConstants.COMPLETE;
}
```
sprinkle a few boundary checks in, of course!

prabu-ssb replied:
then the JSX could be expressed simply as

```
<>
  {isUserStatusComplete(UserStatusKeys.JOURNAL_STATUS) &&
  ...<jsx>
  }
  {!isUserStatusComplete(UserStatusKeys.JOURNAL_STATUS) &&
  ...<jsx>
  }
  {isUserStatusComplete(UserStatusKeys.QUESTIONNAIRE_STATUS) &&
  ...<jsx>
  }
  ...
</>
```

prabu-ssb commented on src/pages/home.tsx:
```diff
@@ -50,7 +51,7 @@ const Home = () => {
     return (
         <Workspace showFooter={true}>
-            {!diaryStatus &&
+            {(!userStatuses ||  userStatuses.surveyStatus !== StatusConstants.COMPLETE) &&
```
see comment regarding `isUserStatusComplete(...)`  below

prabu-ssb commented on src/pages/progress.tsx:
```diff
@@ -1,39 +1,25 @@
-import {FireContext, UserContext} from '../contexts';
-import { useTranslation } from 'react-i18next';
+import React, {useContext, useState} from 'react';
+import {FireContext, UserContext, UserStatusesKeys} from '../contexts';
```
I think we should move the UserStatusKeys to a different place ? Maybe `model/User.ts` ? what do you think ?

eilivjl commented:
Added utility functions to check user statuses in /home and /progress, exchanged update method in useQuestionnaire with Prabus new updateUserStatuses. Update both survey status and journal status on complete. Moved UserStatusKeys to model/user

prabu-ssb commented on src/firebase/model/User.ts:
```diff
@@ -60,3 +67,19 @@ export const getRespondentDetailsSecure = (respondentDetails: RespondentDetails)
     };
 };
+
+const isStatus = (statusKey: UserStatusesKeys, statusValue: StatusConstants): boolean => {
```
When I see this together in one place, I think we can do still better - as it is not clear which status we are talking about (given we have statuses on almost everything!). how about --

 ```
const checkProgress = (statusKey: UserStatusesKeys, statusValue: StatusConstants) => {
  const {userStatuses} = useContext(UserContext);
  return userStatuses && userStatuses[statusKey] === statusValue
};

const isProgressComplete = (statusKey) => 
  checkProgress(statusKey, StatusConstants.COMPLETE);

const isProgressStarted = (statusKey) => 
  checkProgress(statusKey, StatusConstants.STARTED);

const isProgressNotStarted = (statusKey) =>
  !(isProgressStarted(statusKey) || isProgressComplete(statusKey));


/* === Journal === */
const isJournalComplete = () => 
  isProgressComplete(UserStatusesKeys.JOURNAL_STATUS);

const isJournalStarted = () => 
  isProgressStarted(UserStatusesKeys.JOURNAL_STATUS);

const isJournalNotStarted = () => 
  isProgressNotStarted(UserStatusesKeys.JOURNAL_STATUS);
```

prabu-ssb replied:
will have to check if the use of `useContext` is correct in `checkProgress`. react might find it problematic to call from a simple function. 

do test!

eilivjl replied:
Agree to some degree. Maybe my naming convention isnt the best, but I think it is fairly easy to see what status we are talking about based on the key sent into the function. However, this is more readable


Will check context

prabu-ssb commented on src/hocs/useQuestionnaire.tsx:
```diff
@@ -86,7 +86,7 @@ const useQuestionnaire = () => {
         const statusPromises = [
-            questionnaireRef.set({status}, {merge: true}),
+            updateUserStatus(UserStatusesKeys.QUESTIONNAIRE_STATUS, status),
             firestore.doc(getStatusesPathForUser(respondentDetails.respondentId))
```
Aren't we ready to drop this altogether ?
Does'nt `updateUserStatus` do exactly what `firestore.doc().set()` does ?

eilivjl replied:
loks like I deleted the worng thing. The top one there is for setting questionnaire-status at /questionnaire/data ?

prabu-ssb reviewed (approved):
Just check if the updateUserStatus and firestore.set is redundant/duplicate ..

looks very good otherwise! :)

prabu-ssb reviewed (approved):
nice! :)
